src/concurrency/thread.rs

```diff
@@ -118,6 +118,9 @@ pub struct Thread<'mir, 'tcx> {
     /// The virtual call stack.
     stack: Vec<Frame<'mir, 'tcx, Provenance, FrameData<'tcx>>>,
 
+    /// The index of the topmost local frame in `stack`
```

Comment:
Please state exactly what the invariant for this field is. Probably that it always holds the value computed by `reset_top_local`?

As usual, I am concerned about such redundant state and it getting out of sync. Can we have some sanity checks when debug assertions are enabled?

Reply:
I agree that state getting out of sync is a risk here and I don't like it either.

I will of course add some debug assertions, but what do you think about doing some encapsulation around the stack? This is really metadata for that one field in particular, but the problem is that it really has no encapsulation at all because there is a function in the public interface that just returns a mutable reference to the whole field. Is it too much extra complicatiion to refactor the use of `Machine::stack_mut` in rustc to be more clear operations like `push` and `pop` and `get_nth_frame`?

Reply:
I'm always in favor of more encapsulation. :) However, would it help here? We also have the ThreadManager in Miri that has direct access to all the stacks.